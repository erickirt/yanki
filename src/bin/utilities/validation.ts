import { urlToHostAndPort } from '../../lib/utilities/url'

/**
 * Get host and port components of a URL
 *
 * @param url
 * @returns host and port
 * @throws if URL can't be parsed
 */
export function urlToHostAndPortValidated(url: string): { host: string; port: number } {
	const parsedUrl = urlToHostAndPort(url)

	if (parsedUrl === undefined) {
		throw new Error(`Invalid AnkiConnect URL: "${url}"`)
	}

	return parsedUrl
}
