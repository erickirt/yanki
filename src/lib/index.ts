export { type YankiNote } from './model/yanki-note'
export { getNoteFromMarkdown } from './parse/parse'
export { type CleanOptions, cleanNotes, defaultCleanOptions } from './sync/clean'
export { type ListOptions, defaultListOptions, listNotes } from './sync/list'
export { type SyncOptions, defaultSyncOptions, syncFiles, syncNotes } from './sync/sync'
export { hostAndPortToUrl, urlToHostAndPort } from './utilities/string'
