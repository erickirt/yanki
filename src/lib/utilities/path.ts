import { splitAtFirstMatch } from './string'
import isAbsolutePath from '@stdlib/assert-is-absolute-path'
import path from 'path-browserify-esm'
import slash from 'slash'

// Unused...
// function stripLeadingSlash(filePath: string): string {
// 	return filePath.startsWith('/') ? filePath.slice(1) : filePath
// }

/**
 * The browserify polyfill doesn't implement win32 absolute path detection...
 * @param filePath Normalized path
 * @returns
 */
export function isAbsolute(filePath: string): boolean {
	return isAbsolutePath.posix(filePath) || isAbsolutePath.win32(filePath)
}

const RE_WINDOWS_EXTENDED_LENGTH_PATH = /^\\\\\?\\.+/

// Unused
// const RE_WINDOWS_UNC_PATH = /^\\\\[^\\]+\\[^\\]+/

/**
 * Converts all paths to cross-platform 'mixed' style with forward slashes.
 * Warns on unsupported Windows extended length paths.
 * @param filePath
 * @returns normalized path
 */
export function normalize(filePath: string): string {
	if (RE_WINDOWS_EXTENDED_LENGTH_PATH.test(filePath)) {
		console.warn(`Unsupported extended length path detected: ${filePath}`)
		return filePath
	}

	// If (RE_WINDOWS_UNC_PATH.test(filePath)) {
	// 	console.warn(`Unsupported UNC path detected: ${filePath}`)
	// 	return path.normalize(filePath)
	// }

	const basicPath = slash(filePath)
	const normalizedPath = path.normalize(basicPath)

	// Tricky cases where we still want leading './' to distinguish between relative and "named"" paths,
	// otherwise it's stripped by normalization
	if (basicPath.startsWith('./')) {
		return `./${normalizedPath}`
	}

	return normalizedPath
}

/**
 * Special handling for `/absolute-path.md` style links in Obsidian
 * and static site generators, where absolute paths are relative to a base path
 * instead of the volume root.
 *

 *
 * Paths starting with Windows drive letters, while technically absolute, are _not_ prepended with the base:
 * - If no base path is provided, paths are resolved relative to the the provided CWD.
 * - If paths are relative, the base paths are ignored and the CWD is used.
 *
 * All path values are normalized and in 'mixed' platform style.
 */
export function resolveWithBasePath(
	filePath: string,
	options: {
		/** Relative, absolute, or drive-letter absolute path. Normalized and in the 'mixed' platform style. */
		basePath?: string | undefined
		/** Whether to keep prepend the base if the file path already starts with it. Useful for pseudo-idempotence, but will get it wrong in some edge cases with duplicative path segments. Defaults to false. */
		compoundBase?: boolean | undefined
		/** Relative to the volume root. Normalized and in the 'mixed' platform style. */
		cwd: string
	},
): string {
	// Prep options
	const { basePath, compoundBase = false, cwd } = options

	// Validation
	if (basePath !== undefined) {
		if (!isAbsolute(basePath)) {
			console.warn(`Base path "${basePath}" is not absolute`)
		}

		if (!cwd.startsWith(basePath)) {
			console.warn(`CWD "${cwd}" does not start with base path "${basePath}"`)
		}
	}

	if (!isAbsolute(cwd)) {
		console.warn(`CWD "${cwd}" is not absolute`)
	}

	// Absolute
	if (isAbsolute(filePath)) {
		// Path is absolute by drive letter on Windows, or there's not base path to prepend
		if (
			basePath === undefined ||
			/^[A-Za-z]:/.test(filePath) ||
			(!compoundBase && filePath.startsWith(basePath))
		) {
			return filePath
		}

		// Resolve over base
		return path.join(basePath, filePath)
	}

	// Relative
	return path.join(cwd, filePath)
}

/**
 *
 * @param filePath
 * @param cwd Absolute path
 * @returns
 */
export function resolveWithCwd(filePath: string, cwd: string): string {
	if (filePath.startsWith('cwd')) {
		return filePath
	}

	return path.join(cwd, filePath)
}

export function stripBasePath(filePath: string, basePath: string): string {
	const regex = new RegExp(`^${basePath}`, 'i')
	return filePath.replace(regex, '')
}

export function getBaseAndQueryParts(filePath: string): [string, string | undefined] {
	const directoryPath = path.dirname(filePath)
	const fileName = path.basename(filePath)
	const [base, query] = splitAtFirstMatch(fileName, /[#?^]/)
	return [path.join(directoryPath, base), query]
}

export function getBase(filePath: string): string {
	return getBaseAndQueryParts(filePath)[0]
}

export function getQuery(filePath: string): string {
	return getBaseAndQueryParts(filePath).at(1) ?? ''
}

export function hasExtension(filePath: string): boolean {
	return getExtension(filePath) !== ''
}

export function getExtension(filePath: string): string {
	return path.extname(getBase(filePath))
}

export function addExtensionIfMissing(filePath: string, extension: string): string {
	if (hasExtension(filePath)) {
		return filePath
	}

	const [base, query] = getBaseAndQueryParts(filePath)
	return `${base}.${extension}${query ?? ''}`
}

export function quantifyPathDistance(relativePath: string): { down: number; up: number } {
	const up = relativePath.split('../').length - 1
	const down = relativePath.replaceAll('../', '').split('/').length - 1

	return { down, up }
}
