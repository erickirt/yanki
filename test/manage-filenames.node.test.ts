import { globby } from 'globby'
import path from 'path-browserify-esm'
import { expect, it } from 'vitest'
import { syncFiles } from '../src/lib'
import * as pathExtras from '../src/lib/utilities/path'
import { describeWithFileFixture } from './fixtures/file-fixture'
import { cleanUpTempPath, stableResults } from './utilities/stable-sync-results'

describeWithFileFixture(
	'filename management',
	{
		assetPath: './test/assets/test-filename-management/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('uses the prompt as a filename', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'prompt',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const changeMap = new Map<string, string>()
			for (const { filePath, filePathOriginal } of results.synced) {
				expect(filePath).toBeDefined()
				expect(filePathOriginal).toBeDefined()
				changeMap.set(cleanUpTempPath(filePathOriginal)!, cleanUpTempPath(filePath)!)
			}

			expect(changeMap).toMatchInlineSnapshot(`
				Map {
				  "/test-filename-management/case-sensitivity/a.md" => "/test-filename-management/case-sensitivity/A (1).md",
				  "/test-filename-management/case-sensitivity/b.md" => "/test-filename-management/case-sensitivity/A (2).md",
				  "/test-filename-management/case-sensitivity/something-else.md" => "/test-filename-management/case-sensitivity/A (3).md",
				  "/test-filename-management/case-sensitivity/something-lowercase.md" => "/test-filename-management/case-sensitivity/a (4).md",
				  "/test-filename-management/case-sensitivity/something.md" => "/test-filename-management/case-sensitivity/A (5).md",
				  "/test-filename-management/illegal-characters.md" => "/test-filename-management/I might have some deviant filename characters.md",
				  "/test-filename-management/long-content/a.md" => "/test-filename-management/long-content/Lorem ipsum dolor sit amet, consectetur adipiscing elit,... (1).md",
				  "/test-filename-management/long-content/b.md" => "/test-filename-management/long-content/Lorem ipsum dolor sit amet, consectetur adipiscing elit,... (2).md",
				  "/test-filename-management/multi-line/i-have-many-lines.md" => "/test-filename-management/multi-line/This.md",
				  "/test-filename-management/some-question-too.md" => "/test-filename-management/some-question (1).md",
				  "/test-filename-management/some-question.md" => "/test-filename-management/some-question (2).md",
				  "/test-filename-management/something.md" => "/test-filename-management/some-question (3).md",
				  "/test-filename-management/subfolder/some-question-too.md" => "/test-filename-management/subfolder/some-question (1).md",
				  "/test-filename-management/subfolder/some-question.md" => "/test-filename-management/subfolder/some-question (2).md",
				  "/test-filename-management/subfolder/something.md" => "/test-filename-management/subfolder/some-question (3).md",
				  "/test-filename-management/very-long-title-one-word.md" => "/test-filename-management/this note has a very long title indeed what can be done....md",
				  "/test-filename-management/very-long-title.md" => "/test-filename-management/thisnotehasaverylongonewordtitlecanwestillsplititusingthe....md",
				}
			`)

			// Do it again to check for stability
			const tempPath = path.posix.dirname(context.markdownFiles[0])
			const newFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*.md`, {
				absolute: true,
			})
			const newAllFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*`, {
				absolute: true,
			})

			const resultsRound2 = await syncFiles(newFileList, {
				allFilePaths: newAllFileList,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'prompt',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const renamedFilesMap = new Map<string, string>()
			for (const syncedFile of resultsRound2.synced) {
				const original = cleanUpTempPath(syncedFile.filePathOriginal)
				const current = cleanUpTempPath(syncedFile.filePath)
				if (original !== current) {
					renamedFilesMap.set(original!, current!)
				}
			}

			expect(renamedFilesMap).toMatchInlineSnapshot(`Map {}`)
		})
	},
)

describeWithFileFixture(
	'more filename management',
	{
		assetPath: './test/assets/test-filename-management/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('uses the response as a filename', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'response',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const changeMap = new Map<string, string>()
			for (const { filePath, filePathOriginal } of results.synced) {
				expect(filePath).toBeDefined()
				expect(filePathOriginal).toBeDefined()
				changeMap.set(cleanUpTempPath(filePathOriginal)!, cleanUpTempPath(filePath)!)
			}

			expect(changeMap).toMatchInlineSnapshot(`
				Map {
				  "/test-filename-management/case-sensitivity/a.md" => "/test-filename-management/case-sensitivity/B (1).md",
				  "/test-filename-management/case-sensitivity/b.md" => "/test-filename-management/case-sensitivity/B (2).md",
				  "/test-filename-management/case-sensitivity/something-else.md" => "/test-filename-management/case-sensitivity/B (3).md",
				  "/test-filename-management/case-sensitivity/something-lowercase.md" => "/test-filename-management/case-sensitivity/b (4).md",
				  "/test-filename-management/case-sensitivity/something.md" => "/test-filename-management/case-sensitivity/B (5).md",
				  "/test-filename-management/illegal-characters.md" => "/test-filename-management/I might have some deviant filename characters.md",
				  "/test-filename-management/long-content/a.md" => "/test-filename-management/long-content/Jowl sausage, spare ribs meatloaf ham fatback pork... (1).md",
				  "/test-filename-management/long-content/b.md" => "/test-filename-management/long-content/Jowl sausage, spare ribs meatloaf ham fatback pork... (2).md",
				  "/test-filename-management/multi-line/i-have-many-lines.md" => "/test-filename-management/multi-line/And this.md",
				  "/test-filename-management/some-question-too.md" => "/test-filename-management/some-answer (1).md",
				  "/test-filename-management/some-question.md" => "/test-filename-management/some-answer (2).md",
				  "/test-filename-management/something.md" => "/test-filename-management/some-answer (3).md",
				  "/test-filename-management/subfolder/some-question-too.md" => "/test-filename-management/subfolder/some-answer (1).md",
				  "/test-filename-management/subfolder/some-question.md" => "/test-filename-management/subfolder/some-answer (2).md",
				  "/test-filename-management/subfolder/something.md" => "/test-filename-management/subfolder/some-answer (3).md",
				  "/test-filename-management/very-long-title-one-word.md" => "/test-filename-management/this note has a very long title indeed what can be done....md",
				  "/test-filename-management/very-long-title.md" => "/test-filename-management/thisnotehasaverylongonewordtitlecanwestillsplititusingthe....md",
				}
			`)

			// Do it again to check for stability
			const tempPath = path.posix.dirname(context.markdownFiles[0])
			const newFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*.md`, {
				absolute: true,
			})
			const newAllFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*`, {
				absolute: true,
			})

			const resultsRound2 = await syncFiles(newFileList, {
				allFilePaths: newAllFileList,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'response',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const renamedFilesMap = new Map<string, string>()
			for (const syncedFile of resultsRound2.synced) {
				const original = cleanUpTempPath(syncedFile.filePathOriginal)
				const current = cleanUpTempPath(syncedFile.filePath)
				if (original !== current) {
					renamedFilesMap.set(original!, current!)
				}
			}

			expect(renamedFilesMap).toMatchInlineSnapshot(`Map {}`)
		})
	},
)

describeWithFileFixture(
	'even more filename management',
	{
		assetPath: './test/assets/test-filename-management/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('does not touch the filename unless asked', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'off',
				namespace: context.namespace,
			})

			const changeMap = new Map<string, string>()
			for (const { filePath, filePathOriginal } of results.synced) {
				expect(filePath).toBeDefined()
				expect(filePathOriginal).toBeDefined()
				changeMap.set(cleanUpTempPath(filePathOriginal)!, cleanUpTempPath(filePath)!)
			}

			expect(changeMap).toMatchInlineSnapshot(`
				Map {
				  "/test-filename-management/case-sensitivity/a.md" => "/test-filename-management/case-sensitivity/a.md",
				  "/test-filename-management/case-sensitivity/b.md" => "/test-filename-management/case-sensitivity/b.md",
				  "/test-filename-management/case-sensitivity/something-else.md" => "/test-filename-management/case-sensitivity/something-else.md",
				  "/test-filename-management/case-sensitivity/something-lowercase.md" => "/test-filename-management/case-sensitivity/something-lowercase.md",
				  "/test-filename-management/case-sensitivity/something.md" => "/test-filename-management/case-sensitivity/something.md",
				  "/test-filename-management/illegal-characters.md" => "/test-filename-management/illegal-characters.md",
				  "/test-filename-management/long-content/a.md" => "/test-filename-management/long-content/a.md",
				  "/test-filename-management/long-content/b.md" => "/test-filename-management/long-content/b.md",
				  "/test-filename-management/multi-line/i-have-many-lines.md" => "/test-filename-management/multi-line/i-have-many-lines.md",
				  "/test-filename-management/some-question-too.md" => "/test-filename-management/some-question-too.md",
				  "/test-filename-management/some-question.md" => "/test-filename-management/some-question.md",
				  "/test-filename-management/something.md" => "/test-filename-management/something.md",
				  "/test-filename-management/subfolder/some-question-too.md" => "/test-filename-management/subfolder/some-question-too.md",
				  "/test-filename-management/subfolder/some-question.md" => "/test-filename-management/subfolder/some-question.md",
				  "/test-filename-management/subfolder/something.md" => "/test-filename-management/subfolder/something.md",
				  "/test-filename-management/very-long-title-one-word.md" => "/test-filename-management/very-long-title-one-word.md",
				  "/test-filename-management/very-long-title.md" => "/test-filename-management/very-long-title.md",
				}
			`)
		})
	},
)

describeWithFileFixture(
	'complex note type filename management',
	{
		assetPath: './test/assets/test-minimal-notes/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('uses the prompt as a filename across many note types', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'prompt',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const changeMap = new Map<string, string>()
			for (const { filePath, filePathOriginal } of results.synced) {
				expect(filePath).toBeDefined()
				expect(filePathOriginal).toBeDefined()
				changeMap.set(cleanUpTempPath(filePathOriginal)!, cleanUpTempPath(filePath)!)
			}

			expect(changeMap).toMatchInlineSnapshot(`
				Map {
				  "/test-minimal-notes/cloze-with-nothing-else.md" => "/test-minimal-notes/a lonely cloze.md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-back-and-no-front.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (1).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-front-and-no-back.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (2).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line-with-empty-frontmatter.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (3).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line-with-frontmatter.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (4).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (5).md",
				  "/test-minimal-notes/basic-and-reversed-card-with-no-back.md" => "/test-minimal-notes/I'm a question to which there is no answer.md",
				  "/test-minimal-notes/basic-and-reversed-card-with-no-front.md" => "/test-minimal-notes/I'm an answer to which there is no question.md",
				  "/test-minimal-notes/basic-and-reversed-card-with-extra.md" => "/test-minimal-notes/I'm question which is sometimes the answer (1).md",
				  "/test-minimal-notes/basic-and-reversed-card.md" => "/test-minimal-notes/I'm question which is sometimes the answer (2).md",
				  "/test-minimal-notes/basic-with-back-and-no-front-with-empty-frontmatter.md" => "/test-minimal-notes/I'm the back of the card (1).md",
				  "/test-minimal-notes/basic-with-back-and-no-front.md" => "/test-minimal-notes/I'm the back of the card (2).md",
				  "/test-minimal-notes/basic-with-front-and-no-back.md" => "/test-minimal-notes/I'm the front of the card (1).md",
				  "/test-minimal-notes/basic.md" => "/test-minimal-notes/I'm the front of the card (2).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-empty-frontmatter.md" => "/test-minimal-notes/I'm the prompt (1).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-frontmatter.md" => "/test-minimal-notes/I'm the prompt (2).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-multiple-emphasis-and-ignored-answer-style.md" => "/test-minimal-notes/I'm the prompt (3).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-multiple-emphasis.md" => "/test-minimal-notes/I'm the prompt (4).md",
				  "/test-minimal-notes/basic-type-in-the-answer.md" => "/test-minimal-notes/I'm the prompt (5).md",
				  "/test-minimal-notes/basic-with-front-and-cloze-like-back.md" => "/test-minimal-notes/I'm the question.md",
				  "/test-minimal-notes/cloze-with-no-preamble.md" => "/test-minimal-notes/is the.md",
				  "/test-minimal-notes/basic-with-empty-frontmatter.md" => "/test-minimal-notes/My frontmatter is empty.md",
				  "/test-minimal-notes/cloze-with-extra-empty.md" => "/test-minimal-notes/This card has a (1).md",
				  "/test-minimal-notes/cloze-with-extra.md" => "/test-minimal-notes/This card has a (2).md",
				  "/test-minimal-notes/cloze-with-style.md" => "/test-minimal-notes/This card has a (3).md",
				  "/test-minimal-notes/cloze.md" => "/test-minimal-notes/This card has a (4).md",
				  "/test-minimal-notes/basic-with-cloze-like-back-and-no-front.md" => "/test-minimal-notes/This looks a lot like a cloze but it's a basic answer.md",
				  "/test-minimal-notes/basic-with-empty-everything.md" => "/test-minimal-notes/Untitled (1).md",
				  "/test-minimal-notes/basic-with-front-image-markdown-embed-and-no-back.md" => "/test-minimal-notes/Untitled (2).md",
				  "/test-minimal-notes/basic-with-front-image-wiki-embed-and-no-back.md" => "/test-minimal-notes/Untitled (3).md",
				}
			`)

			// Do it again to check for stability
			const tempPath = path.posix.dirname(context.markdownFiles[0])
			const newFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*.md`, {
				absolute: true,
			})
			const newAllFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*`, {
				absolute: true,
			})

			const resultsRound2 = await syncFiles(newFileList, {
				allFilePaths: newAllFileList,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'prompt',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const renamedFilesMap = new Map<string, string>()
			for (const syncedFile of resultsRound2.synced) {
				const original = cleanUpTempPath(syncedFile.filePathOriginal)
				const current = cleanUpTempPath(syncedFile.filePath)
				if (original !== current) {
					renamedFilesMap.set(original!, current!)
				}
			}

			expect(renamedFilesMap).toMatchInlineSnapshot(`Map {}`)
		})
	},
)

describeWithFileFixture(
	'more complex note type filename management',
	{
		assetPath: './test/assets/test-minimal-notes/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('uses the response as a filename across many note types', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'response',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const changeMap = new Map<string, string>()
			for (const { filePath, filePathOriginal } of results.synced) {
				expect(filePath).toBeDefined()
				expect(filePathOriginal).toBeDefined()
				changeMap.set(cleanUpTempPath(filePathOriginal)!, cleanUpTempPath(filePath)!)
			}

			expect(changeMap).toMatchInlineSnapshot(`
				Map {
				  "/test-minimal-notes/cloze-with-nothing-else.md" => "/test-minimal-notes/a lonely cloze.md",
				  "/test-minimal-notes/cloze-with-extra-empty.md" => "/test-minimal-notes/cloze (1).md",
				  "/test-minimal-notes/cloze-with-extra.md" => "/test-minimal-notes/cloze (2).md",
				  "/test-minimal-notes/cloze-with-no-preamble.md" => "/test-minimal-notes/cloze (3).md",
				  "/test-minimal-notes/cloze.md" => "/test-minimal-notes/cloze (4).md",
				  "/test-minimal-notes/cloze-with-style.md" => "/test-minimal-notes/emphasized but un-hinted cloze.md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-back-and-no-front.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (1).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-front-and-no-back.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (2).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line-with-empty-frontmatter.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (3).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line-with-frontmatter.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (4).md",
				  "/test-minimal-notes/basic-with-type-in-the-answer-like-single-line.md" => "/test-minimal-notes/I look a lot like the thing you need to type in, but i'm... (5).md",
				  "/test-minimal-notes/basic-and-reversed-card-with-no-back.md" => "/test-minimal-notes/I'm a question to which there is no answer.md",
				  "/test-minimal-notes/basic-and-reversed-card-with-no-front.md" => "/test-minimal-notes/I'm an answer to which there is no question.md",
				  "/test-minimal-notes/basic-and-reversed-card-with-extra.md" => "/test-minimal-notes/I'm an answer which is sometimes the question (1).md",
				  "/test-minimal-notes/basic-and-reversed-card.md" => "/test-minimal-notes/I'm an answer which is sometimes the question (2).md",
				  "/test-minimal-notes/basic-with-back-and-no-front-with-empty-frontmatter.md" => "/test-minimal-notes/I'm the back of the card (1).md",
				  "/test-minimal-notes/basic-with-back-and-no-front.md" => "/test-minimal-notes/I'm the back of the card (2).md",
				  "/test-minimal-notes/basic.md" => "/test-minimal-notes/I'm the back of the card (3).md",
				  "/test-minimal-notes/basic-with-front-and-no-back.md" => "/test-minimal-notes/I'm the front of the card.md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-empty-frontmatter.md" => "/test-minimal-notes/I'm the thing you need to type on the card (1).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-frontmatter.md" => "/test-minimal-notes/I'm the thing you need to type on the card (2).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-multiple-emphasis-and-ignored-answer-style.md" => "/test-minimal-notes/I'm the thing you need to type on the card (3).md",
				  "/test-minimal-notes/basic-type-in-the-answer-with-multiple-emphasis.md" => "/test-minimal-notes/I'm the thing you need to type on the card (4).md",
				  "/test-minimal-notes/basic-type-in-the-answer.md" => "/test-minimal-notes/I'm the thing you need to type on the card (5).md",
				  "/test-minimal-notes/basic-with-empty-frontmatter.md" => "/test-minimal-notes/My frontmatter is empty.md",
				  "/test-minimal-notes/basic-with-cloze-like-back-and-no-front.md" => "/test-minimal-notes/This looks a lot like a cloze but it's a basic answer.md",
				  "/test-minimal-notes/basic-with-front-and-cloze-like-back.md" => "/test-minimal-notes/This looks a lot like a cloze or two here's a hint.md",
				  "/test-minimal-notes/basic-with-empty-everything.md" => "/test-minimal-notes/Untitled (1).md",
				  "/test-minimal-notes/basic-with-front-image-markdown-embed-and-no-back.md" => "/test-minimal-notes/Untitled (2).md",
				  "/test-minimal-notes/basic-with-front-image-wiki-embed-and-no-back.md" => "/test-minimal-notes/Untitled (3).md",
				}
			`)

			// Do it again to check for stability
			const tempPath = path.posix.dirname(context.markdownFiles[0])
			const newFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*.md`, {
				absolute: true,
			})
			const newAllFileList = await globby(`${pathExtras.normalize(tempPath)}/**/*`, {
				absolute: true,
			})

			const resultsRound2 = await syncFiles(newFileList, {
				allFilePaths: newAllFileList,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'response',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			const renamedFilesMap = new Map<string, string>()
			for (const syncedFile of resultsRound2.synced) {
				const original = cleanUpTempPath(syncedFile.filePathOriginal)
				const current = cleanUpTempPath(syncedFile.filePath)
				if (original !== current) {
					renamedFilesMap.set(original!, current!)
				}
			}

			expect(renamedFilesMap).toMatchInlineSnapshot(`Map {}`)
		})
	},
)

describeWithFileFixture(
	'filename collision management',
	{
		assetPath: './test/assets/test-filename-management-collision/',
		cleanUpAnki: true,
		cleanUpTempFiles: true,
	},
	(context) => {
		it('handles filename management deck collisions', async () => {
			const results = await syncFiles(context.markdownFiles, {
				allFilePaths: context.allFiles,
				ankiConnectOptions: {
					autoLaunch: true,
				},
				ankiWeb: false,
				dryRun: false,
				manageFilenames: 'prompt',
				namespace: context.namespace,
				syncMediaAssets: 'off',
			})

			expect(stableResults(results)).toMatchSnapshot()
		})
	},
)
