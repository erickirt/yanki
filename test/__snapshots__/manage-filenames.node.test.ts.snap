// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filename collision management > handles filename management deck collisions 1`] = `
{
  "ankiWeb": false,
  "deletedDecks": [],
  "deletedMedia": [],
  "dryRun": false,
  "duration": 0,
  "fixedDatabase": false,
  "namespace": "Yanki Test - ab27c9c22de22d6b",
  "synced": [
    {
      "action": "created",
      "filePath": "/test-filename-management-collision/deck a/deck a.md",
      "filePathOriginal": "/test-filename-management-collision/deck a/basic.md",
      "note": {
        "deckName": "deck a",
        "fields": {
          "Back": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH back model-yanki-basic"><p>the back</p></div>",
          "Front": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH front model-yanki-basic"><p>deck a</p></div>",
          "YankiNamespace": "Yanki Test - ab27c9c22de22d6b",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "/test-filename-management-collision/deck b.md/deck b.md.md",
      "filePathOriginal": "/test-filename-management-collision/deck b.md/basic md.md",
      "note": {
        "deckName": "deck b.md",
        "fields": {
          "Back": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH back model-yanki-basic"><p>the back</p></div>",
          "Front": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH front model-yanki-basic"><p>deck b.md</p></div>",
          "YankiNamespace": "Yanki Test - ab27c9c22de22d6b",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "/test-filename-management-collision/deck b.md/deck b.md",
      "filePathOriginal": "/test-filename-management-collision/deck b.md/basic.md",
      "note": {
        "deckName": "deck b.md",
        "fields": {
          "Back": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH back model-yanki-basic"><p>the back</p></div>",
          "Front": "<!-- This HTML was generated by Yanki, a Markdown to Anki converter. Do not edit directly. --><div class="yanki namespace-yanki-test-HASH front model-yanki-basic"><p>deck b</p></div>",
          "YankiNamespace": "Yanki Test - ab27c9c22de22d6b",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
  ],
}
`;
